### PR TITLE
Support a stable double backward on linalg.det for real inputs

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #80217
* #80074

The complex case still fails. I do not know why.

Fixes https://github.com/pytorch/pytorch/issues/62327
Fixes https://github.com/pytorch/pytorch/issues/53364

### CODE REVIEW

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 1]
```diff
@@ -4050,14 +4050,61 @@ Tensor linalg_matrix_exp_differential(
...
+  // the mask is true and f2 to the tensors for which the mask is false This
+  // function is used when we have a formula that works for, say, all
+  // non-singular inputs and another one for when the inputs are singular. See
+  // for example det_backward
+  auto n_true = mask.count_nonzero().item<int64_t>();
+  if (n_true == mask.numel()) {
+    return f1(t, ts...);
+  } else if (n_true == 0) {
+    return f2(t, ts...);
+  } else {
```
Comment (nikitaved) [inline torch/csrc/autograd/FunctionsManual.cpp:4072]: Maybe we can skip these branches as to avoid syncs?

Comment (Lezcano) [reply torch/csrc/autograd/FunctionsManual.cpp:4072]: It's how it was implemented before. I think it's fine.

Comment (nikitaved) [reply torch/csrc/autograd/FunctionsManual.cpp:4072]: Yeah, but we can make it without any syncs whatsoever. Could you please provide some benchmarks to see whether the juice is wroth the squeeze?

Comment (nikitaved) [reply torch/csrc/autograd/FunctionsManual.cpp:4072]: OK, I see that the previous implementation was syncing as well. Let's add a note/todo/comment about this?

Comment (Lezcano) [reply torch/csrc/autograd/FunctionsManual.cpp:4072]: will do

Comment (nikitaved) [reply torch/csrc/autograd/FunctionsManual.cpp:4072]: But I still believe that it is very likely that most use cases of `det` are full-rank matrices, maybe with a small subset of singular matrices :)

Comment (Lezcano) [reply torch/csrc/autograd/FunctionsManual.cpp:4072]: That is true. And that's the reason for this sync, to completely avoid having to call the singular part at all.

**Review by nikitaved — APPROVED** [review]
LGTM!

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 2]
```diff
@@ -4050,14 +4050,69 @@ Tensor linalg_matrix_exp_differential(
...
+    // Do
+    // ret = t.new_empty(t.shape)
+    // ret[mask] = f1(t1[mask], ..., tn[mask])
+    // ret[~mask] = f2(t1[~mask], ..., tn[~mask])
+    auto not_mask = mask.logical_not();
+    return t.new_empty(t.sizes())
+        .index_put_({mask}, f1(t_masked, ts.index({mask})...))
+        .index_put_(
+            {not_mask}, f2(t.index({not_mask}), ts.index({not_mask})...));
```
Comment (nikitaved) [inline torch/csrc/autograd/FunctionsManual.cpp:4085]: nit: second `index_put_` could use in-place not I guess?

**Review by albanD — APPROVED** [review]
:+1:

File: aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp [hunk 3]
```diff
@@ -1614,7 +1614,12 @@ static void lu_factor(const Tensor& input, const Tensor& pivots, const Tensor& i
   const auto preferred_backend = at::globalContext().linalgPreferredBackend();
 #ifdef USE_CUSOLVER
   const auto lu_factor_cusolver = [batch_size, m, n](const Tensor& input, const Tensor& pivots, const Tensor& infos, bool compute_pivots) {
-    if (batch_size == 1 || m != n || m >= 512) {
+    // In CUDA 10.2, lu_factor_looped_cusolver does not finish the computations when the input
+    // matrix is exactly singular. The returned pivots contain garbage. This breaks linalg.det
+    // Now, batched_cublas does not handle rectangular matrices, so we still dispatch to
+    // looped_cusolver even if m != n.
+    constexpr bool looped_correct = CUSOLVER_VERSION >= 11100;
+    if (m != n || (looped_correct && (batch_size == 1 || m >= 512))) {
```
Comment (Lezcano) [inline aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp:1622]: I went for this fix to minimise the skips @IvanYashchuk

File: aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp [hunk 4]
```diff
@@ -1614,7 +1614,12 @@ static void lu_factor(const Tensor& input, const Tensor& pivots, const Tensor& i
   const auto preferred_backend = at::globalContext().linalgPreferredBackend();
 #ifdef USE_CUSOLVER
   const auto lu_factor_cusolver = [batch_size, m, n](const Tensor& input, const Tensor& pivots, const Tensor& infos, bool compute_pivots) {
-    if (batch_size == 1 || m != n || m >= 512) {
+    // In CUDA 10.2, lu_factor_looped_cusolver does not finish the computations when the input
+    // matrix is exactly singular. The returned pivots contain garbage. This breaks linalg.det
+    // Now, batched_cublas does not handle rectangular matrices, so we still dispatch to
+    // looped_cusolver even if m != n.
+    constexpr bool looped_correct = CUSOLVER_VERSION >= 11100;
```
Comment (IvanYashchuk) [inline aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp:1621]: This corresponds to the CUDA Toolkit version 11.2.1 (cusolver and cuda versions do not match) is that what you wanted? In newer versions, pivots are good and do not contain garbage?

Comment (Lezcano) [reply aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp:1621]: Yep, in newer versions this bug is corrected (tested locally, but also the CI now passes).

**Review by malfet — APPROVED** [review]

**Pull request closed without merging**